### PR TITLE
CDAP-9026 Cleanup entity roles on entity deletion

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: github.com/caskdata/cdap

See issues for details.

### CODE REVIEW

File: cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java [hunk 1]
```diff
@@ -200,15 +197,54 @@ private synchronized void performGroupBasedGrant(EntityId entityId, Principal pr
...
+    final String curEntityRolePrefix = Joiner.on(ENTITY_ROLE_PREFIX).join("", entityId.toString());
+
+    Predicate<Role> filter = new Predicate<Role>() {
+      public boolean apply(Role role) {
+        return role.getName().startsWith(curEntityRolePrefix);
+      }
+    };
+
+    // get all roles and filter roles which belongs to this this entity
+    Set<Role> existingRoles = listAllRoles();
```
Comment (Nishith) [inline cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java]: Can replace this with 
`return Sets.newHashSet(Collections2.filter(listAllRoles(), filter));`

File: cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java [hunk 2]
```diff
@@ -200,15 +197,54 @@ private synchronized void performGroupBasedGrant(EntityId entityId, Principal pr
...
+  private void cleanUpEntityRole(Role entityRole) {
+    // this should not be called for any other role except entity roles i.e. the ones which start with
+    // ENTITY_ROLE_PREFIX
+    if (!entityRole.getName().startsWith(ENTITY_ROLE_PREFIX)) {
+      throw new IllegalArgumentException(String.format("The given role %s is not an entity role. " +
+                                                         "Please use drop role to remove this role.", entityRole));
+    }
+    // if there are privileges associated with this role then don't clean it up
+    if (!listPrivileges(entityRole).isEmpty()) {
+      return;
```
Comment (Nishith) [inline cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java]: Add a debug log?

File: cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java [hunk 3]
```diff
@@ -200,15 +197,54 @@ private synchronized void performGroupBasedGrant(EntityId entityId, Principal pr
...
+    }
+    try {
       binding.dropRole(entityRole);
+      LOG.debug("Successfully dropped role {}", entityRole);
+    } catch (RoleNotFoundException e) {
+      // This is a dot role. It should be ok for deletion to fail. This happens because while creating a new entity,
+      // we first revoke any orphaned privileges on the entity. During that operation this role may not exist.
+      LOG.debug("Trying to delete role {}, but it was not found. Ignoring since its an entity role.", entityRole);
```
Comment (Nishith) [inline cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java]: s/its/it's/

**Review by Nishith — CHANGES_REQUESTED** [review]
A few minor comments, please take a look.

**Comment by rohitsinha54** [comment]
@Nishith : Addressed comments. Also look at  second commit called fixes which has some fixes for related issues found during testing on cluster.

**Review by Nishith — APPROVED** [review]

**Pull request merged**